File: lib/mix/tasks/catalog_sync.ex
defmodule Mix.Tasks.Catalog.Sync do
  @moduledoc """
  Syncs agent catalog from an external GitHub repo.

  ## Usage

      mix catalog.sync [--repo URL] [--clear]

  ## Options

    * `--repo` - GitHub repo URL (default: https://github.com/msitarzewski/agency-agents)
    * `--clear` - Remove existing catalog before sync
  """
  @shortdoc "Sync agent catalog from external repo"

  use Mix.Task

  @default_repo "https://github.com/msitarzewski/agency-agents"
  @skip_dirs ~w(.github integrations examples .git)

  @impl Mix.Task
  def run(args) do
    {opts, _, _} =
      OptionParser.parse(args, strict: [repo: :string, clear: :boolean])

    repo = opts[:repo] || @default_repo
    catalog_dir = Path.join(File.cwd!(), "priv/catalog")

    if opts[:clear] do
      Mix.shell().info("Clearing existing catalog...")
      File.rm_rf!(catalog_dir)
    end

    tmp_dir = Path.join(System.tmp_dir!(), "catalog_sync_#{:erlang.unique_integer([:positive])}")

    try do
      Mix.shell().info("Cloning #{repo}...")

      case System.cmd("git", ["clone", "--depth", "1", repo, tmp_dir], stderr_to_stdout: true) do
        {_, 0} -> :ok
        {output, _} -> Mix.raise("Failed to clone repo: #{output}")
      end

      {agents_count, categories} = process_repo(tmp_dir, catalog_dir)
      write_categories(catalog_dir, categories)

      Mix.shell().info("Synced #{agents_count} agents across #{length(categories)} categories")
    after
      File.rm_rf!(tmp_dir)
    end
  end

  defp process_repo(repo_dir, catalog_dir) do
    entries = File.ls!(repo_dir)

    categories =
      entries
      |> Enum.filter(fn entry ->
        path = Path.join(repo_dir, entry)
        File.dir?(path) and entry not in @skip_dirs
      end)
      |> Enum.filter(fn dir ->
        path = Path.join(repo_dir, dir)
        path |> File.ls!() |> Enum.any?(&String.ends_with?(&1, ".md"))
      end)

    agents_count =
      categories
      |> Enum.map(fn category ->
        category_path = Path.join(repo_dir, category)
        out_dir = Path.join([catalog_dir, "agents", category])
        File.mkdir_p!(out_dir)

        category_path
        |> File.ls!()
        |> Enum.filter(&String.ends_with?(&1, ".md"))
        |> Enum.reject(&(&1 == "README.md"))
        |> Enum.count(fn file ->
          content = File.read!(Path.join(category_path, file))
          {frontmatter, body} = parse_frontmatter(content)

          if frontmatter["name"] do
            yaml_map = build_agent_yaml(frontmatter, body, category)
            slug = yaml_map["name"]
            yaml_content = encode_agent_yaml(yaml_map)
            File.write!(Path.join(out_dir, "#{slug}.yaml"), yaml_content)
            true
          else
            false
          end
        end)
      end)
      |> Enum.sum()

    category_maps =
      categories
      |> Enum.map(fn cat ->
        %{
          "id" => cat,
          "name" =>
            cat
            |> String.replace("-", " ")
            |> String.split()
            |> Enum.map_join(" ", &String.capitalize/1),
          "description" => "#{String.capitalize(cat)} agents"
        }
      end)

    {agents_count, category_maps}
  end

  defp write_categories(catalog_dir, categories) do
    File.mkdir_p!(catalog_dir)
    yaml_content = encode_categories_yaml(categories)
    File.write!(Path.join(catalog_dir, "categories.yaml"), yaml_content)
  end

  # Manual YAML encoding (yaml_elixir is read-only)

  defp encode_agent_yaml(map) do
    [
      "name: #{map["name"]}",
      "display_name: #{yaml_quote(map["display_name"])}",
      "description: #{yaml_quote(map["description"])}",
      "category: #{map["category"]}",
      "emoji: #{yaml_quote(map["emoji"])}",
      "tags: #{encode_tags(map["tags"])}",
      "harness: #{map["harness"]}",
      "model: #{map["model"]}",
      "system_prompt: |",
      indent(map["system_prompt"] || "", "  ")
    ]
    |> Enum.join("\n")
    |> Kernel.<>("\n")
  end

  defp encode_categories_yaml(categories) do
    categories
    |> Enum.map(fn cat ->
      "- id: #{cat["id"]}\n  name: #{yaml_quote(cat["name"])}\n  description: #{yaml_quote(cat["description"])}"
    end)
    |> Enum.join("\n")
    |> Kernel.<>("\n")
  end

  defp yaml_quote(nil), do: ~s("")
  defp yaml_quote(""), do: ~s("")

  defp yaml_quote(s) do
    ~s("#{String.replace(s, "\"", "\\\"")}")
  end

  defp encode_tags(nil), do: "[]"
  defp encode_tags([]), do: "[]"

  defp encode_tags(tags) when is_list(tags) do
    items = Enum.map_join(tags, ", ", &to_string/1)
    "[#{items}]"
  end

  defp encode_tags(_), do: "[]"

  defp indent(text, prefix) do
    text
    |> String.split("\n")
    |> Enum.map_join("\n", &(prefix <> &1))
  end

  @doc "Parses YAML frontmatter from markdown content. Returns {frontmatter_map, body_string}."
  def parse_frontmatter(content) do
    case String.split(String.trim_leading(content), "---", parts: 3) do
      ["", frontmatter, body] ->
        case YamlElixir.read_from_string(frontmatter) do
          {:ok, parsed} -> {parsed, String.trim(body)}
          {:error, _} -> {%{}, String.trim(content)}
        end

      _ ->
        {%{}, String.trim(content)}
    end
  end

  @doc "Converts a display name to a URL-safe slug."
  def slugify(name) do
    name
    |> String.downcase()
    |> String.replace(~r/[^a-z0-9\s-]/, "-")
    |> String.replace(~r/\s+/, "-")
    |> String.replace(~r/-+/, "-")
    |> String.trim("-")
  end

  @doc "Builds a catalog agent YAML map from parsed frontmatter and markdown body."
  def build_agent_yaml(frontmatter, body, category) do
    display_name = frontmatter["name"] || ""

    %{
      "name" => slugify(display_name),
      "display_name" => display_name,
      "description" => frontmatter["description"] || "",
      "category" => category,
      "emoji" => frontmatter["emoji"] || "",
      "tags" => [],
      "harness" => "claude_code",
      "model" => "claude-sonnet-4-6",
      "system_prompt" => body
    }
  end
end

File: src/scripts/catalog-sync.ts
#!/usr/bin/env bun
/**
 * Syncs agent catalog from an external GitHub repo.
 *
 * Usage:
 *   bun run src/scripts/catalog-sync.ts [--repo URL] [--clear]
 */
import { mkdirSync, rmSync, readdirSync, statSync, readFileSync, writeFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { tmpdir } from "os";
import { safeYamlLoad } from "../utils/yaml";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DEFAULT_REPO = "https://github.com/msitarzewski/agency-agents";
const SKIP_DIRS = new Set([".github", "integrations", "examples", ".git"]);
const CATALOG_DIR = join(__dirname, "..", "..", "catalog");

function parseArgs(): { repo: string; clear: boolean } {
  const args = process.argv.slice(2);
  let repo = DEFAULT_REPO;
  let clear = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--repo" && args[i + 1]) {
      repo = args[++i];
    } else if (args[i] === "--clear") {
      clear = true;
    }
  }

  return { repo, clear };
}

function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, "-")
    .replace(/\s+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

function parseFrontmatter(content: string): { frontmatter: Record<string, unknown>; body: string } {
  const trimmed = content.trimStart();
  const parts = trimmed.split("---");
  if (parts.length >= 3 && parts[0] === "") {
    try {
      const frontmatter = safeYamlLoad(parts[1]) as Record<string, unknown>;
      const body = parts.slice(2).join("---").trim();
      return { frontmatter: frontmatter ?? {}, body };
    } catch {
      return { frontmatter: {}, body: trimmed };
    }
  }
  return { frontmatter: {}, body: trimmed };
}

function buildAgentYaml(
  frontmatter: Record<string, unknown>,
  body: string,
  category: string,
): Record<string, unknown> {
  const displayName = String(frontmatter.name ?? "");
  return {
    name: slugify(displayName),
    display_name: displayName,
    description: String(frontmatter.description ?? ""),
    category,
    emoji: String(frontmatter.emoji ?? ""),
    tags: [],
    harness: "claude_code",
    model: "claude-sonnet-4-6",
    system_prompt: body,
  };
}

function encodeAgentYaml(map: Record<string, unknown>): string {
  const q = (s: unknown) => {
    const str = String(s ?? "");
    return str ? `"${str.replace(/"/g, '\\"')}"` : '""';
  };

  const lines = [
    `name: ${map.name}`,
    `display_name: ${q(map.display_name)}`,
    `description: ${q(map.description)}`,
    `category: ${map.category}`,
    `emoji: ${q(map.emoji)}`,
    `tags: []`,
    `harness: ${map.harness}`,
    `model: ${map.model}`,
    `system_prompt: |`,
    ...String(map.system_prompt ?? "")
      .split("\n")
      .map((line) => `  ${line}`),
  ];
  return lines.join("\n") + "\n";
}

async function main() {
  const { repo, clear } = parseArgs();

  if (clear) {
    console.log("Clearing existing catalog...");
    rmSync(CATALOG_DIR, { recursive: true, force: true });
  }

  const tmpDir = join(tmpdir(), `catalog_sync_${Date.now()}`);

  try {
    console.log(`Cloning ${repo}...`);
    const result = Bun.spawnSync(["git", "clone", "--depth", "1", repo, tmpDir]);
    if (result.exitCode !== 0) {
      console.error(`Failed to clone repo: ${new TextDecoder().decode(result.stderr)}`);
      process.exit(1);
    }

    // Find category dirs (dirs containing .md files, excluding skip list)
    const entries = readdirSync(tmpDir);
    const categories = entries.filter((entry) => {
      if (SKIP_DIRS.has(entry)) return false;
      const path = join(tmpDir, entry);
      if (!statSync(path).isDirectory()) return false;
      return readdirSync(path).some((f) => f.endsWith(".md"));
    });

    let agentsCount = 0;

    for (const category of categories) {
      const categoryPath = join(tmpDir, category);
      const outDir = join(CATALOG_DIR, "agents", category);
      mkdirSync(outDir, { recursive: true });

      const mdFiles = readdirSync(categoryPath).filter(
        (f) => f.endsWith(".md") && f !== "README.md",
      );

      for (const file of mdFiles) {
        const content = readFileSync(join(categoryPath, file), "utf-8");
        const { frontmatter, body } = parseFrontmatter(content);

        if (frontmatter.name) {
          const agentYaml = buildAgentYaml(frontmatter, body, category);
          const yamlContent = encodeAgentYaml(agentYaml);
          writeFileSync(join(outDir, `${agentYaml.name}.yaml`), yamlContent);
          agentsCount++;
        }
      }
    }

    // Write categories.yaml
    mkdirSync(CATALOG_DIR, { recursive: true });
    const categoryMaps = categories.map((cat) => ({
      id: cat,
      name: cat
        .replace(/-/g, " ")
        .split(" ")
        .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
        .join(" "),
      description: `${cat.charAt(0).toUpperCase() + cat.slice(1)} agents`,
    }));

    const categoriesYaml =
      categoryMaps
        .map((c) => `- id: ${c.id}\n  name: "${c.name}"\n  description: "${c.description}"`)
        .join("\n") + "\n";

    writeFileSync(join(CATALOG_DIR, "categories.yaml"), categoriesYaml);

    console.log(`Synced ${agentsCount} agents across ${categories.length} categories`);
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

main();
